lib/request.js

```diff
@@ -23,6 +24,100 @@ function refreshAuth(type, config) {
   });
 }
 
+const execChannels = [
+  'stdin',
+  'stdout',
+  'stderr',
+  'error',
+  'resize'
+];
+
+/**
+ * Determine whether a failed Kubernetes API response is asking for an upgrade
+ * @param {object} body - response body object from Kubernetes
+ * @property {string} status - request status
+ * @property {number} code - previous request's response code
+ * @property {message} message - previous request response message
+ * @returns {boolean} Upgrade the request
+ */
+
+function isUpgradeRequired(body) {
+  return body.status === 'Failure'
+    && body.code === 400
+    && body.message === 'Upgrade request required';
+}
+
+/**
+ * Construct the query parameters necessary for performing an execAction
+ * @param {object} qs - query parameter object as passed by user
+ * @property {string[]} command - (required) command and arguments to be executed
+ * @property {string} container - (optional) container in which to execute the command; optional if there's only one container in the pod
+ * @property {boolean} tty - (optional) whether to hold the websocket connection open for an interactive session (not yet implemented)
+ * @param {function} cb - callback to handle the response
+ * @returns {string} All necessary query parameters as a string ready for inclusion in URL
+ */
+function buildExecQueryParams(qs, cb) {
+  const command = qs.command;
+  if (!command || !Array.isArray(command)) {
+    return cb(new Error(`expected command to be an array of strings, not: ${JSON.stringify(command)}`));
+  }
+
+  // The command query parameter gets specified multiple times to
+  // indicate arguments to the initial binary
+  // (command=ls&command=-al), so it cannot be handled in an object
+  // like the rest of the qs.
+  const commandAsParams = command.map((cmd) => `command=${encodeURIComponent(cmd)}`);
+  delete qs.command;
+
+  const execParams = Object.assign(qs, {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    // TODO: allow users to have interactive sessions
+    tty: false
+  });
+
+  return Object.keys(execParams)
+    .map((key) => `${key}=${execParams[key]}`)
+    .concat(commandAsParams)
+    .join('&');
+}
+
+/**
+ * Upgrade a request into a Websocket transaction & process the result
+ * @param {ApiRequestOptions} options - Options object
+ * @param {callback} cb - The callback that handles the response
+ */
+
+function upgradeRequest(options, cb) {
+  const qs = buildExecQueryParams(options.qs, cb);
+
+  const wsUrl = `${options.baseUrl}/${options.uri}?${qs}`;
+  const protocol = 'base64.channel.k8s.io';
+  const ws = new WebSocket(wsUrl, protocol, options);
+
+  const messages = [];
+  ws.on('message', (msg) => {
+    const channel = execChannels[msg.slice(0, 1)];
+    const message = Buffer.from(msg.slice(1), 'base64').toString('ascii');
+    messages.push({ channel, message });
+  });
+
+  ws.on('error', (err) => {
+    err.messages = messages;
+    cb(err, messages);
+  });
+
+  ws.on('close', (code, reason) => cb(null, {
```

Comment:
I wasn't sure what the response format for this should be. The joined command output is available on `res.body`, but I figure it may also be useful to see the messages, websocket exit code, and reason (if any).

Reply:
yeah, i'm not sure either.  I like smaller incremental PRs/changes.  So one thing we could do is try what you have here, but don't advertise support yet, and expect to iterate on the exact result format, etc.